java/ql/src/Security/CWE/CWE-346/UnvalidatedCors.ql

```diff
@@ -0,0 +1,55 @@
+/**
+ * @name Cors header being set from remote source
+ * @description Cors header is being set from remote source, allowing to control the origin.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id java/unvalidated-cors-origin-set
+ * @tags security
+ *       external/cwe/cwe-346
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.frameworks.Servlets
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+// Check for Access-Control-Allow-Credentials as well, this ensures fair chances of exploitability.
+predicate satisfyAllowCredentials(MethodAccess header, MethodAccess check) {
```

Comment:
Probably cleaner to remove the `check` and apply the in-same-callable constraint at the callsite.